sound/soc/sof/control.c

```diff
@@ -65,6 +65,37 @@ static inline u32 ipc_to_mixer(u32 value, u32 *volume_map, int size)
 	return i - 1;
 }
 
+static void snd_sof_refresh_control(struct snd_sof_control *scontrol)
+{
+	struct sof_ipc_ctrl_data *cdata = scontrol->control_data;
+	struct snd_soc_component *scomp = scontrol->scomp;
+	enum sof_ipc_ctrl_type ctrl_type;
+	int ret;
+
+	if (!scontrol->updated_in_firmware)
+		return;
+
+	if (!pm_runtime_active(scomp->dev))
+		return;
+
+	if (scontrol->cmd == SOF_CTRL_CMD_BINARY)
+		ctrl_type = SOF_IPC_COMP_GET_DATA;
+	else
+		ctrl_type = SOF_IPC_COMP_GET_VALUE;
+
+	/* set the ABI header values */
+	cdata->data->magic = SOF_ABI_MAGIC;
+	cdata->data->abi = SOF_ABI_VERSION;
+	/* get all the component data from DSP */
+	ret = snd_sof_ipc_set_get_comp_data(scontrol, ctrl_type,
+					    SOF_CTRL_TYPE_VALUE_CHAN_GET,
+					    scontrol->cmd, false);
+	if (ret < 0)
+		dev_err(scomp->dev, "error: failed to get control data: %d\n", ret);
+
+	scontrol->updated_in_firmware = false;
```

Comment:
if there a possibility of a race condition, e.g. with the DSP sending two IPCs. If the refresh_control deals with the first one and clears this flag after it's set by the second IPC, it could be that the second update is lost? the test on line 75 would prevent the second access to the register.

Reply:
The chances are pretty low that it would happen, but in theory it is a possibility.
The IPC tx/rx is lock protected so the IPC is serialized, if the whole RX processing happens between snd_sof_ipc_set_get_comp_data() and we clear the flag then we can miss one update.

It can be mitigated by moving the flag clear before snd_sof_ipc_set_get_comp_data(), with that we might force a second read via IPC in the worst case but we will not miss updates.

Thank you for pointing this out!